### DISCUSSION TITLE
TypeScript type error when using `@babel/template`'s `template()` function

### DISCUSSION THREAD

**Discussion by liushuai0109** [babel/babel#13361] [Q&A]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

I was trying to use @babel/template to do something like a babel plugin.
The code bellow was copied from Babel's doc page https://babeljs.io/docs/en/babel-template, but it didn't work
```js
  const buildRequire = template(`
    var %%importName%% = require(%%source%%);
  `);

  const ast = buildRequire({
    importName: t.identifier('myModule'),
    source: t.stringLiteral('my-module'),
  });

  console.log(generate(ast).code);
```
I tried to run the code above in a TypeScript project, but it appeared to be something wrong.

here is my package.json dependencies

```json
{
  "dependencies": {
    "@babel/parser": "^7.14.3",
    "@babel/template": "^7.12.13",
    "@babel/traverse": "^7.14.2",
    "gulp": "^4.0.2",
    "lodash": "^4.17.21",
    "nodegit": "^0.27.0",
    "through2": "^4.0.2",
    "vinyl": "^2.2.1"
  },
  "devDependencies": {
    "@babel/types": "^7.14.2",
    "@tencent/eslint-config-tencent": "^0.13.0",
    "@types/gulp": "^4.0.8",
    "@types/jest": "^26.0.23",
    "@types/lodash": "^4.14.169",
    "@types/node": "^15.0.2",
    "@types/nodegit": "^0.27.2",
    "@types/through2": "^2.0.36",
    "@types/vinyl": "^2.0.4",
    "@typescript-eslint/eslint-plugin": "^4.23.0",
    "@typescript-eslint/parser": "^4.23.0",
    "commander": "^7.2.0",
    "eslint": "^7.26.0",
    "jest": "^26.6.3",
    "ts-jest": "^26.5.6",
    "ts-node": "^9.1.1",
    "typedoc": "^0.20.36",
    "typescript": "^4.2.4"
  }
}
```

### Current and expected behavior

it seems to be something wrong with Types, here is the error messages.

TS2345: Argument of type 'Statement | Statement[]' is not assignable to parameter of type 'Node'.   Type 'Statement[]' is not assignable to type 'Node'.     Type 'Statement[]' is missing the following properties from type 'VariableDeclarator': type, id, leadingComments, innerComments, and 4 more.

i

### Environment

Babel version: 7.14.3
Node: 14.16.1
npm: 6.14.12
OS: macOS 11.13.1
Monorepo: no

**Comment by nicolo-ribaudo** [comment]
The `template` function can return either a single node or multiple nodes. For example, this code would throw a runtime error because `ast` is an array:
```js
  const buildRequire = template(`
    var %%importName%% = require(%%source%%);
    buildRequire++;
  `);

  const ast = buildRequire({
    importName: t.identifier('myModule'),
    source: t.stringLiteral('my-module'),
  });

  console.log(generate(ast).code);
```

You can either cast the `buildRequire` result:
```js
  const buildRequire = template(`
    var %%importName%% = require(%%source%%);
  `);

  const ast = buildRequire({
    importName: t.identifier('myModule'),
    source: t.stringLiteral('my-module'),
  }) as t.Statement;

  console.log(generate(ast).code);
```

Or use the `template.statement` function, that never returns an array:
```js
  const buildRequire = template.statement(`
    var %%importName%% = require(%%source%%);
  `);

  const ast = buildRequire({
    importName: t.identifier('myModule'),
    source: t.stringLiteral('my-module'),
  });

  console.log(generate(ast).code);
```
